pkg/kwok/metrics/histogram.go

```diff
@@ -0,0 +1,181 @@
+	Name      string
+
+	// Help provides information about this Histogram.
+	//
+	// Metrics with the same fully-qualified name must have the same Help
+	// string.
+	Help string
+
+	// ConstLabels are used to attach fixed labels to this metric. Metrics
+	// with the same fully-qualified name must have the same label names in
+	// their ConstLabels.
+	//
+	// ConstLabels are only used rarely. In particular, do not use them to
+	// attach the same labels to all your metrics. Those use cases are
+	// better covered by target labels set by the scraping Prometheus
+	// server, or by one specific metric (e.g. a build_info or a
+	// machine_role metric). See also
+	// https://prometheus.io/docs/instrumenting/writing_exporters/#target-labels-not-static-scraped-labels
+	ConstLabels prometheus.Labels
+	Buckets     []float64
+}
+
+type bucketCounter struct {
+	bucket  float64
+	counter atomic.Int64
+}
+
+// Histogram is custom type emulating prometheus.Histogram
+type Histogram struct {
+	desc *prometheus.Desc
+
+	buckets []*bucketCounter
+
+	observations atomic.Int64
+
+	sumMu sync.Mutex
+	sum   float64
+}
+
+// NewHistogram creates new Histogram based on Histogram options
+func NewHistogram(opts HistogramOpts) *Histogram {
+	desc := prometheus.NewDesc(
+		prometheus.BuildFQName(opts.Namespace, opts.Subsystem, opts.Name),
+		opts.Help,
+		nil,
+		opts.ConstLabels,
+	)
+
+	buckets := make([]*bucketCounter, 0, len(opts.Buckets))
+	for _, b := range opts.Buckets {
+		buckets = append(buckets, &bucketCounter{
+			bucket: b,
+		})
+	}
+
+	sort.Slice(buckets, func(i, j int) bool {
+		return buckets[i].bucket < buckets[j].bucket
+	})
+
+	his := &Histogram{
+		desc:    desc,
+		buckets: buckets,
+	}
+	prometheus.NewHistogram(prometheus.HistogramOpts{})
+	return his
+}
+
+// Desc returns prometheus.Desc used by every Prometheus Metric.
+func (h *Histogram) Desc() *prometheus.Desc {
+	return h.desc
+}
+
+// Write writes out Histogram data to the Metric dto.
+func (h *Histogram) Write(out *dto.Metric) error {
+	buckets := make([]*dto.Bucket, 0, len(h.buckets))
+
+	var cumCount uint64
+	for _, bucket := range h.buckets {
+		cumCount += uint64(bucket.counter.Load())
+		buckets = append(
+			buckets,
+			&dto.Bucket{
+				CumulativeCount: proto.Uint64(cumCount),
+				UpperBound:      proto.Float64(bucket.bucket),
+			},
+		)
+	}
+
+	h.sumMu.Lock()
+	sum := h.sum
+	h.sumMu.Unlock()
+	his := &dto.Histogram{
+		Bucket:      buckets,
+		SampleCount: proto.Uint64(uint64(h.observations.Load())),
+		SampleSum:   proto.Float64(sum),
+	}
+
+	out.Histogram = his
+
+	return nil
+}
+
+// Describe sends metric description to a channel.
+func (h *Histogram) Describe(ch chan<- *prometheus.Desc) {
+	ch <- h.desc
+}
+
+// Collect sends histogram to a prometheus Metric channel.
+func (h *Histogram) Collect(ch chan<- prometheus.Metric) {
+	ch <- h
+}
+
+// Set is used to increment value at le bucket and add val to the sum.
+func (h *Histogram) Set(le string, val float64) (err error) {
```

Comment:
I tested it and it didn't behave as I expected, I refactored it and now it meets expectations 😄.

https://github.com/wzshiming/kwok/commit/e4f49e3acf3ae0915cb3898e8f9cbf326888aa93

`git apply <(curl https://github.com/wzshiming/kwok/commit/e4f49e3acf3ae0915cb3898e8f9cbf326888aa93.patch)`

Reply:
Amazing, sorry I could not make it work right away, I did not exactly know how it should behave :smile: 
I tried applying patch but it produced an error
Do you mind pushing on the branch :smile:

Reply:
BTY, I added [auto docs for API](https://kwok.sigs.k8s.io/docs/generated/apis/), so this PR needs to rebase then exec `./hack/update-api-docs.sh`

Reply:
Rebased and executed
I'll apply the patch by hand. git shows error (patch does not apply)